c/serverStatusService.c

```diff
@@ -0,0 +1,284 @@
+  setDefaultJSONRESTHeaders(response);
+  writeHeader(response);
+  jsonStart(out);
+  if(ctime_r(&ltime, tstamp) != NULL){
+    if(tstamp[strlen(tstamp) - 1] != '\0'){
+      tstamp[strlen(tstamp) - 1] = '\0';
+    }
+    jsonAddString(out, "timestamp", tstamp);
+  }
+  if(getenv("ZSS_LOG_FILE") != NULL){
+    jsonAddString(out, "logDirectory", getenv("ZSS_LOG_FILE"));
+  } else {
+    jsonAddString(out, "logDirectory", "ZSS_LOG_FILE not defined in environment variables");
+  }
+  jsonAddString(out, "agentName", "zss");
+  jsonAddString(out, "agentVersion", context->productVersion);
+  jsonAddString(out, "arch", unameRet.sysname);
+  jsonAddString(out, "osRelease", unameRet.release);
+  jsonAddString(out, "osVersion", unameRet.version);
+  jsonAddString(out, "hardwareIdentifier", unameRet.machine);
+  jsonAddString(out, "hostname", hostnameBuffer);
+  jsonAddString(out, "nodename", unameRet.nodename);
+  jsonStartObject(out, "userEnvironment");
+  char *envVar = *environ;
+  for (int i = 1; envVar; i++) {
+    char *equalSign = strchr(envVar, '=');
+    if(equalSign == NULL){
+      break;
+    }
+    int nameLen = strlen(envVar) - strlen(equalSign);
+    int nameBufferLen = nameLen + 1;
+    char *name = safeMalloc(nameBufferLen, "env_var name");
+    if(name == NULL){
+      break;
+    }
+    memcpy(name, envVar, nameLen);
+    name[nameLen] = '\0';
+    jsonAddString(out, name, equalSign+1);
+    safeFree(name, nameBufferLen);
+    envVar = *(environ+i);
+  }
+  jsonEndObject(out);
+  jsonAddInt(out, "demandPagingRate", demandPaging);
+  jsonAddInt(out, "stdCP_CPU_Util", cpuUtil);
+  jsonAddInt(out, "stdCP_MVS_SRM_CPU_Util", mvsSrm);
+  jsonAddInt(out, "ZAAP_CPU_Util", zaapUtil);
+  jsonAddInt(out, "ZIIP_CPU_Util", ziipUtil);
+  jsonAddInt(out, "PID", getpid());
+  jsonAddInt(out, "PPID", getppid());
+  jsonEnd(out);
+  finishResponse(response);
+}
+
+static int serveStatus(HttpService *service, HttpResponse *response) {
+  HttpRequest *request = response->request;
+  ServerAgentContext *context = service->userPointer;
+  //This service is conditional on RBAC being enabled because it is a 
+  //sensitive URL that only RBAC authorized users should be able to access
+  JsonObject *dataserviceAuth = jsonObjectGetObject(context->serverConfig, "dataserviceAuthentication");
+  int rbacParm = jsonObjectGetBoolean(dataserviceAuth, "rbac");
+  if(!rbacParm){
+     respondWithError(response, HTTP_STATUS_BAD_REQUEST, "Set dataserviceAuthentication.rbac to true in server configuration");
+  } else {
+    if(!strcmp(request->method, methodGET)){
+      char *l1 = stringListPrint(request->parsedFile, 2, 1, "/", 0);
+      if(!strcmp(l1, "")){
+        respondWithServerRoutes(response);
+      }else if (!strcmp(l1, "config")){
+        respondWithServerConfig(response, context->serverConfig);
+      }else if (!strcmp(l1, "log")) {
+        char* logDir = getenv("ZSS_LOG_FILE");
+        if(logDir == NULL || strcmp(logDir, "")){
```

Comment:
Seems like a bug. Did you mean the following?
```c
if (logDir != NULL && strcmp(logDir, "")) {
```

Reply:
You can flatten these kind of bugs by using macros or (preferably) inline functions instead of ```strcmp()``` for checking string equality. I prefer using bool functions as I dislike using boolean logic for ```strcmp()``` which is almost always used for equality;

````
inline bool strne(const char * lhs, const char * rhs) { 
  return lhs != NULL && rhs != NULL && strcmp(lhs, rhs) != 0;
}
````

Reply:
I quite like this inline function, I think I will implement it!

Reply:
For that particular case you may prefer a different function.

````c
inline bool str_is_empty(const char * str) {
  return str ==  NULL || str[0] == '\0';
}
````
These kind of string functions are as old as the hills. C has a poor API for checking string equality which necessitates writing wrappers. You can see examples in lots of [open source software](https://github.com/systemd/systemd/blob/master/src/basic/string-util.h).